internal/controller/controllers/agentserviceconfig_controller_test.go

```diff
@@ -0,0 +1,189 @@
+package controllers
+
+import (
+	"context"
+
+	"github.com/golang/mock/gomock"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	routev1 "github.com/openshift/api/route/v1"
+	aiv1beta1 "github.com/openshift/assisted-service/internal/controller/api/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes/scheme"
+	ctrl "sigs.k8s.io/controller-runtime"
+	fakeclient "sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+const (
+	testName                         = "agent"
+	testAgentServiceConfigKind       = "testKind"
+	testAgentServiceConfigAPIVersion = "testAPIVersion"
+)
+
+func newTestReconciler(initObjs ...runtime.Object) *AgentServiceConfigReconciler {
+	c := fakeclient.NewFakeClientWithScheme(scheme.Scheme, initObjs...)
+	return &AgentServiceConfigReconciler{
+		Client:    c,
+		Scheme:    scheme.Scheme,
+		Log:       ctrl.Log.WithName("testLog"),
+		Namespace: testNamespace,
+	}
+}
+
+var _ = Describe("agentserviceconfig_controller reconcile", func() {
+	var (
+		asc      *aiv1beta1.AgentServiceConfig
+		ascr     *AgentServiceConfigReconciler
+		ctx      = context.Background()
+		mockCtrl *gomock.Controller
+		route    *routev1.Route
+	)
+
+	BeforeEach(func() {
+		mockCtrl = gomock.NewController(GinkgoT())
+		asc = newDefaultAgentServiceConfig()
+		ascr = newTestReconciler(asc)
+
+		// AgentServiceConfig created route is missing Host.
+		// We create one here with a value set for Host so that
+		// reconcile does not fail.
+		route, _ = ascr.newAgentRoute(asc)
+		route.Spec.Host = "testHost"
+		Expect(ascr.Client.Create(ctx, route)).To(BeNil())
+	})
+
+	AfterEach(func() {
+		mockCtrl.Finish()
+		Expect(ascr.Client.Delete(ctx, asc)).To(Succeed())
+		Expect(ascr.Client.Delete(ctx, route)).To(Succeed())
+	})
+
+	It("reconcile should succeed", func() {
+		result, err := ascr.Reconcile(newAgentServiceConfigRequest(asc))
+		Expect(err).To(BeNil())
+		Expect(result).To(Equal(ctrl.Result{}))
+	})
+})
+
+var _ = Describe("ensureAgentLocalAuthSecret", func() {
+	var (
+		asc             *aiv1beta1.AgentServiceConfig
+		ascr            *AgentServiceConfigReconciler
+		ctx             = context.Background()
+		mockCtrl        *gomock.Controller
+		privateKey      = "test-private-key"
+		publicKey       = "test-public-key"
+		localAuthSecret *corev1.Secret
+	)
+
+	BeforeEach(func() {
+		mockCtrl = gomock.NewController(GinkgoT())
+		asc = newDefaultAgentServiceConfig()
+		ascr = newTestReconciler(asc)
+	})
+
+	AfterEach(func() {
+		mockCtrl.Finish()
+		Expect(ascr.Client.Delete(ctx, asc)).To(Succeed())
+	})
+
+	Context("with an existing local auth secret", func() {
```

Comment:
I personally like @carbonin recommendation to simply call the `ensure*` functions directly. That would allow you to, in this PR to specifcially test the auth secret functionality, in a follow-up we can cover the remainder of `ensure*` functions and maybe cover the rest with a few handcrafted scenarios where calling `Reconcile(someRequest)` makes sense.

Reply:
I think that's what we're doing in this context. The top one is the only one that calls `Reconcile` which I already mentioned probably should have been a separate PR.